doc/read-the-docs-site/adr/0005-inlining-saturated-functions.md

```diff
@@ -0,0 +1,185 @@
+# ADR 5: Inlining fully saturated functions
+
+Date: 2022-12
+
+## Authors
+
+Marty Stumpf <marty.stumpf@iohk.io>
+
+## Status
+
+Accepted
+
+## Context
+
+As part of our effort to increase script capacity,
+we take from [Plutonomy](https://github.com/well-typed/plutonomy),
+which gives 15-30% speedup on benchmarks.
+See [PLT-1044](https://input-output.atlassian.net/browse/PLT-1044).
+
+The goal is to integrate what we know already works there and
+also add optimizations we can do before UPLC (Plutonomy only applies optimization to UPLC).
+We currently have an inliner pass for PIR and UPLC.
+One of the optimization we can add is inlining of saturated functions in both PIR and UPLC.
+From our analysis, this is one of the optimizations that contributes to the speedup in Plutonomy.
+
+For example:
+
+```haskell
+let f x y = 1
+    y = f q
+in f a b 
+```
+
+we should inline `f` on line 3 (even though it’s used more than once), because it’s saturated,
+i.e. it has all its arguments, and we can reduce it.
+([PLT-1146](https://input-output.atlassian.net/browse/PLT-1146))
+
+To do this we need to count the number of type/term lambda abstractions,
+so we know how many term/type arguments they have.
+We also need to decide whether or not to inline based on the call site, which we don’t currently.
+
+## Decision
+
+For a first round of implementation, we are adding a simplified pass.
+We will refine the implementation over time.
+In each round, we benchmark and test and investigate the results.
+
+### The heuristic: Counting lambdas and arguments
+
+In this inline pass we check for a binding that is a lambda abstraction and is fully applied in the body.
+
+E.g.,
+
+```haskell
+let v = rhs in body
+```
+
+We pattern match the _rhs_ with `LamAbs` or `TyAbs` (lambda abstraction for terms or types),
+and count the number of lambdas.
+Then, in the _body_, we check for term or type application (`Apply` or `TyInst`) of _v_.
+
+If _v_ is fully applied in the body, i.e., if
+
+1. the number of type lambdas equals the number of type arguments applied, and
+2. the number of term lambdas equals the number of term arguments applied, and
+
+if other call site inlining conditions are satisfied, (we currently inline too little, this will be improved later, see below)
+
+we inline _v_, i.e., replace its occurrence with _rhs_ in the _body_.
+For the rest of the discussion here we focus on the conditions 1 and 2 only.
+
+E.g., for
+
+```haskell
+let f = \x.\y.x+y in
+    f 2 4
+```
+
+the _body_ `f 2 4` is a function application of `f`. Since `f` is fully applied, we inline `f`.
+
+### Examples
+
+Below are some examples that involve a _body_ that is not fully reducible/applied but following our heuristic is beneficial.
+
+#### Example 1: function in body
+
+```haskell
+let f = \x . x
+     g = \y . f
+in g a
+```
+
+`f` and `g` each has 1 lambda. However, `g`'s _body_ includes `f` which also has a lambda.
```

Comment:
I'm confused, why is this example not "fully reducible/applied"? I don't think we have a clear concept of what that means? This is the stuff I'd really like us to be clear about: what _exactly_ do we mean by "saturated", "fully applied" etc. There are several options!